Add GraphModule.recompile_submodules, use for regional inductor

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166002
* #165996

This makes it so that `GraphModule.recompile()` will also recompile any submodules that are also graph modules, which allows us to pass all existing regional inductor tests without skipping. 

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

jamesjwu committed:
Update

[ghstack-poisoned]

anijain2305 commented on torch/fx/graph_module.py:
```diff
@@ -848,6 +848,17 @@ def call_wrapped(self, *args, **kwargs):
 
         return python_code
 
+    def recompile_submodules(self) -> list[tuple[str, PythonCode]]:
```
I think this should be done automatically for `recompile` method

jamesjwu committed:
Update on "Add GraphModule.recompile_submodules, use for regional inductor"




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

jamesjwu committed:
Update on "Add GraphModule.recompile_submodules, use for regional inductor"




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

jamesjwu committed:
Rebase on "Add GraphModule.recompile_submodules, use for regional inductor"




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

jamesjwu committed:
Update on "Add GraphModule.recompile_submodules, use for regional inductor"


This makes it so that `GraphModule.recompile()` will also recompile any submodules that are also graph modules, which allows us to pass all existing regional inductor tests without skipping. 

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

jamesjwu committed:
Update on "Add GraphModule.recompile_submodules, use for regional inductor"


This makes it so that `GraphModule.recompile()` will also recompile any submodules that are also graph modules, which allows us to pass all existing regional inductor tests without skipping. 

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
